scripts/check_lint.py

```diff
@@ -0,0 +1,232 @@
+    '-readability/casting',
+
+    # Objective-C needs use type 'long' for interop between types like NSInteger
+    # and printf-style functions.
+    '-runtime/int',
+
+    # cpplint is generally confused by Objective-C mixing with C++.
+    #   * Objective-C method invocations in a for loop make it think its a
+    #     range-for
+    #   * Objective-C dictionary literals confuse brace spacing
+    #   * Empty category declarations ("@interface Foo ()") look like function
+    #     invocations
+    '-whitespace',
+]
+
+_CPPLINT_OBJC_OPTIONS = [
+    # cpplint normally excludes Objective-C++
+    '--extensions=h,m,mm',
+
+    # Objective-C style allows longer lines
+    '--linelength=100',
+
+    '--filter=' + ','.join(_CPPLINT_OBJC_FILTERS),
+]
+
+
+def main():
+  global _dry_run
+
+  parser = argparse.ArgumentParser(description='Lint source files.')
+  parser.add_argument('--dry-run', '-n', action='store_true',
+                      help='Show what the linter would do without doing it')
+  parser.add_argument('--all', action='store_true',
+                      help='run the linter over all known sources')
+  parser.add_argument('rev_or_files', nargs='*',
+                      help='A single revision that specifies a point in time '
+                           'from which to look for changes. Defaults to '
+                           'origin/master. Alternatively, a list of specific '
+                           'files or git pathspecs to lint.')
+  args = command_trace.parse_args(parser)
+
+  if args.dry_run:
+    _dry_run = True
+    command_trace.enable_tracing()
+
+  pool = checker.Pool()
+
+  sources = _unique(source.CC_DIRS + source.OBJC_DIRS + source.PYTHON_DIRS)
+  patterns = git.make_patterns(sources)
+
+  files = git.find_changed_or_files(args.all, args.rev_or_files, patterns)
+  check(pool, files)
+
+  pool.exit()
+
+
+def check(pool, files):
+  group = source.categorize_files(files)
+
+  for kind, files in group.kinds.items():
+    for chunk in checker.shard(files):
+      if not chunk:
+        continue
+
+      linter = _linters[kind]
+      pool.submit(linter, chunk)
+
+
+def lint_cc(files):
+  return _run_cpplint([], files)
+
+
+def lint_objc(files):
+  return _run_cpplint(_CPPLINT_OBJC_OPTIONS, files)
+
+
+def _run_cpplint(options, files):
+  scripts_dir = os.path.dirname(os.path.abspath(__file__))
+  cpplint = os.path.join(scripts_dir, 'cpplint.py')
+
+  command = [sys.executable, cpplint, '--quiet']
+  command.extend(options)
+  command.extend(files)
+
+  return _read_output(command)
+
+
+_flake8_warned = False
+
+
+def lint_py(files):
+  flake8 = which('flake8')
+  if flake8 is None:
+    global _flake8_warned
+    if not _flake8_warned:
+      _flake8_warned = True
+      _logger.warn(textwrap.dedent(
+          """
+          Could not find flake8 on the path; skipping python lint.
+          Install with:
+
+            pip install --user flake8
+          """))
+    return
+
+  command = [sys.executable, flake8]
```

Comment:
I just tried it out on my system, and this doesn't work quite right:

```
$ /usr/bin/python /usr/bin/flake8 scripts/check_lint.py 
Traceback (most recent call last):
  File "/usr/bin/flake8", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3088, in <module>
    @_call_aside
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3072, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3101, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 574, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 892, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 778, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'flake8==3.5.0' distribution was not found and is required by the application
$
```

This is fine though:
```
$ /usr/bin/flake8 scripts/check_lint.py
$
```

The issue is that (on my linux system) /usr/bin/python is python2, whereas flake8 is a python3 package/binary.

Since the `which` method verifies that the flake8 "binary" is executable, can we simply drop sys.executable from the command? (That *does* work on my system.)

Reply:
OK. We trust flake8 to pick its own interpreter.

I've left cpplint alone since it's not marked executable and it's coming from this repo so it can't really know which python it's supposed to run against.